src/ussd_clinic_rapidpro.js

```diff
@@ -140,16 +140,77 @@ go.app = function() {
                         );
                     }
                 },
-                next: "state_get_contact"
+                next: "state_whatsapp_check"
             });
         });
 
+        self.add("state_whatsapp_contact_check", function(name, opts) {
+            var msisdn = utils.normalize_msisdn(
+                _.get(self.im.user.answers, "state_enter_msisdn", self.im.user.addr), "ZA");
+            self.whatsapp.contact_check(msisdn, false).then(_.noop, _.noop);
+            return self.whatsapp.contact_check(msisdn, true)
+                .then(function(result) {
+                    self.im.user.set_answer("on_whatsapp", result);
+                    return self.states.create("state_clinic_code");
+                }).catch(function(e) {
+                    // Go to error state after 3 failed HTTP requests
+                    opts.http_error_count = _.get(opts, "http_error_count", 0) + 1;
+                    if(opts.http_error_count === 3) {
+                        self.im.log.error(e.message);
+                        return self.states.create("__error__", {return_state: name});
+                    }
+                    return self.states.create(name, opts);
+                });
+        });
+
+        self.add("state_clinic_code", function(name, opts) {
+            var text;
+            if(!self.im.user.answers.on_whatsapp) {
```

Comment:
In `state_start`, if they choose the `yes` option, they go straight here without going through `state_whatsapp_contact_check`, so this field won't be set, and they'll always end up on `state_not_on_whatsapp`